docs/design_proposals/configurable-kubecontext.md

```diff
@@ -0,0 +1,183 @@
+
+### Detailed discussion
+#### Option in `skaffold.yaml`
+A configuration option in `skaffold.yaml` has the advantage of being most discoverable:
+it is in the place where users configure all aspects of their pipeline.
+In addition, it allows to define the kubecontext per Skaffold profile.
+
+However, it also has questionable implications:
+
+- `skaffold.yaml` is meant to be shared, but kubecontext names vary across users.
+  Sharing therefore makes only sense in a corporate setting where context names are the same across many users.
+  There is however a risk of abuse in settings where sharing the context name does not make sense, for example in open source projects.
+- Due to profile activation by Skaffold profiles, there can be confusing/surprising situations.
+  For example, the current context is `minikube` and activates some profile.
+  This profile is deploying to a different kubecontext `gke_abc_123`. This can be surprising to the user.
+  A solution could be to forbid specifying a kubecontext in a profile if it is activated by a kubecontext (and validate that).
+
+A natural place for the config in `skaffold.yaml` is in `latest.DeployConfig`, resulting in a json path `deploy.kubeContext`.
+
+#### Option in global Skaffold config
+Specifying a default kubecontext globally is straightforward. For example, via new config option
+```yaml
+global:
+  default-context: docker-for-desktop
+```
+
+On the other hand, building a relation between projects and kubecontext needs to solve two questions:
+
+1. How to identify projects
+2. How to save the project/kubecontext relation
+
+##### How to identify projects
+
+There are at least three possibilities:
+
+- Identify projects by their absolute host path.
+  This is guaranteed to be unique, but may break if a user moves his project to another location.
+- Identify projects by a new `metadata.name` entry in `skaffold.yaml` (see also [#2200](https://github.com/GoogleContainerTools/skaffold/issues/2200)).
+  This has the drawback of being potentially not unique, so that users accidentally pin the kubecontext for more projects than intended.
+  On the other hand, this is the standard approach taken by kubernetes resources.
+- Identify project by their initial commit.
+  This variant is stable against relocations.
+  It is also unique unless a user forks a project and wants to define different kubecontexts for each fork.
+  Compared to the other alternatives, it is rather slow.
+
+**\<What option has the best tradeoffs?\>**
+
+Resolution: We will go ahead with the `metadata.name` approach. As the name may not be unique, this requires special documentation.
+
+
+##### How to save project/kubecontext relations
+
+Currently, the Skaffold config uses the kubecontext as identifier.
+
+There are two possibilities to add the relation:
+
+- Reverse the mapping project/kubecontext and save as list under `kube-context` entries:
+  ```yaml
+  kubecontexts:
+  - kube-context: my-context
+    projects:
```

Comment:
@balopat Please correct me if I'm wrong, but my impression was that this variant was the preferred one.

Also, we should carefully think about how to name the new field, as the global config is not versioned and field renaming will be much harder. @tejal29 confused it with GCB projects, and this may happen to other users too. We could make it clearer that this ties the kubecontext to a specific named `skaffold.yaml`, and call the field `skaffoldConfigs` instead.

Reply:
+1 on `skaffoldConfigs`